cpp/oneapi/dal/table/backend/csr_kernels.cpp

```diff
@@ -0,0 +1,226 @@
+}
+
+template <typename Policy, typename BlockData>
+void pull_data_impl(const Policy& policy,
+                    const csr_info& origin_info,
+                    const array<byte_t>& origin_data,
+                    const bool same_type,
+                    const std::int64_t origin_offset,
+                    const std::int64_t block_size,
+                    detail::sparse_block<BlockData>& block,
+                    alloc_kind kind,
+                    bool preserve_mutability) {
+    constexpr std::int64_t block_dtype_size = sizeof(BlockData);
+    constexpr data_type block_dtype = detail::make_data_type<BlockData>();
+
+    const auto origin_dtype_size = detail::get_data_type_size(origin_info.dtype_);
+
+    if (same_type && !alloc_kind_requires_copy(get_alloc_kind(origin_data), kind)) {
+        refer_origin_data(origin_data,
+                          origin_offset * block_dtype_size,
+                          block_size,
+                          block.data,
+                          preserve_mutability);
+    }
+    else {
+        if (block.data.get_count() < block_size || !block.data.has_mutable_data() ||
+            alloc_kind_requires_copy(get_alloc_kind(block.data), kind)) {
+            reset_array(policy, block.data, block_size, kind);
+        }
+
+        auto src_data = origin_data.get_data() + origin_offset * origin_dtype_size;
+        auto dst_data = block.data.get_mutable_data();
+
+        backend::convert_vector(policy,
+                                src_data + origin_offset * block_dtype_size,
+                                dst_data,
+                                origin_info.dtype_,
+                                block_dtype,
+                                block_size);
+    }
+}
+
+template <typename Policy, typename BlockData>
+void pull_column_indices_impl(const Policy& policy,
+                              const array<std::int64_t>& origin_column_indices,
+                              const std::int64_t origin_offset,
+                              const std::int64_t block_size,
+                              detail::sparse_block<BlockData>& block,
+                              alloc_kind kind,
+                              bool preserve_mutability) {
+    if (!alloc_kind_requires_copy(get_alloc_kind(origin_column_indices), kind)) {
+        reset_array(policy, block.column_indices, block_size, kind);
+    }
+    refer_origin_data(origin_column_indices,
+                      origin_offset,
+                      block_size,
+                      block.column_indices,
+                      preserve_mutability);
+}
+
+template <typename Policy, typename BlockData>
+void pull_row_indices_impl(const Policy& policy,
+                           const array<std::int64_t>& origin_row_indices,
+                           const block_info& block_info,
+                           detail::sparse_block<BlockData>& block,
```

Comment:
seems like you need only row_indices from the block in this function. Can you pass only what you actially needed instead of the whole block?